README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Overview
========

The `multiGSEA` package was built to facilitate the use of gene sets in the analysis of high throughput genomics data (primarily RNA-seq). Analysts can orchestrate any number of GSEA methods across a specific contrast using the unified interface provided by the `multiGSEA` function, and a shiny application is provided that facilitates the exploration and interpration of GSEA results.

-   The `multiGSEA` function is a wrapper that orchestrates the execution of any number of user-specified gene set enrichment analyses (GSEA) over a particular experimental contrast of interest. This will create a `MultiGSEAResult` object which stores the results of each GSEA method internally, allowing for easy query and retrieval.
-   A sister `multiGSEA.shiny` package provides an `explore` function, which is invoked on `MultiGSEAREsult` objects returned from a call to `multiGSEA`. The shiny application facilitates interactive exploration of these GSEA results. This application can also be deployed to a shiny server and can be initialized by uploading a serialized `MultiGSEAResult` `*.rds` file.

Full details about using this software package is provided in the included vignette, however a brief description is outlined below.

Installation
============

The multiGSEA suite of package will soon be submitted to bioconductor and installable via the recommended `biocLite` mechanism. In the meantime, these packages can be installed like so:

``` r
# install.packages("devtools")
devtools::install_github("lianos/multiGSEA")
devtools::install_github("lianos/multiGSEA.shiny")
devtools::install_github("lianos/GeneSetDb.MSigDB.Hsapiens.v61")
devtools::install_github("lianos/GeneSetDb.MSigDB.Mmusculs.v61")
```

Example Usage
=============

A subset of the RNA-seq data tumor/nomral samples in the BRCA indication from the TCGA are provided in this package. We will use that data to perform a "camera" and "fry" gene set enrichment analysis of tumor vs normal samples using the MSigDB hallmark and c2 gene set collections with `multiGSEA`.

``` r
library(multiGSEA)
library(dplyr)
gdb <- getMSigGeneSetDb(c('h', 'c2'), 'human')
vm <- exampleExpressionSet(dataset='tumor-vs-normal', do.voom=TRUE)
mg <- multiGSEA(gdb, vm, vm$design, "tumor", methods=c("camera", "fry"))
```

We can view the top "camera" results with the smallest pvalues like so:

``` r
results(mg, "camera") %>% 
  arrange(pval) %>% 
  select(collection, name, padj) %>% 
  head
#>   collection                                        name         padj
#> 1         c2      SOTIRIOU_BREAST_CANCER_GRADE_1_VS_3_UP 5.374230e-46
#> 2         c2 ROSTY_CERVICAL_CANCER_PROLIFERATION_CLUSTER 2.323940e-42
#> 3         c2         NAKAYAMA_SOFT_TISSUE_TUMORS_PCA2_DN 1.333271e-30
#> 4         c2               CROONQUIST_IL6_DEPRIVATION_DN 7.857976e-27
#> 5         c2                     BENPORATH_PROLIFERATION 1.113013e-26
#> 6         c2              KANG_DOXORUBICIN_RESISTANCE_UP 2.298134e-25
```

The shift in expression of the genes within the top gene set can be visualized with the `iplot` function below. This plot produces interactive graphics, but rasteriaed versions are save for use with this `README.md` file:

``` r
iplot(mg, 'c2', 'SOTIRIOU_BREAST_CANCER_GRADE_1_VS_3_UP', type="density")
```

<img src="vignettes/images/README_iplot_density.png" />

``` r
iplot(mg, 'c2', 'SOTIRIOU_BREAST_CANCER_GRADE_1_VS_3_UP', type="boxplot")
```

<img src="vignettes/images/README_iplot_boxplot.png" />


NEWS
CHANGES IN VERSION 0.10
-----------------------

USER VISIBLE CHANGES:

  o MSigGeneSetDb updated to v6.1 and data refactored out into GeneSetDb.* packages
  o `geneSet,MultiGSEAResult` method does not remove duplicated columns present
    in both `geneSet(gdb)` and `logFC`, but rather appends a `*.gs` suffix to
    the duplicated columns from `geneSet`. This happened because I defined
    a geneset with a `logFC` column (from the experiment the set was derived)
    from, but this collided with (and superseded) the `logFC` value for the gene
    returned from `logFC(mgresult)`, and messed up interpretation of a geneset's
    move across a contrast.
  o Introduces idea of gene- and geneset-level metadata in a GeneSetDb. The
    former (gene-level metadata) was always possible by creating a GeneSetDb
    with a data.frame which has more columns than necessary, ie. instead of
    just having a collection,name,featureId data.frame, you could add a symbol
    column. This would simply be part of the GeneSetDb@db data.table. If there
    are columns in your input data.frame that are all the same value per
    collection,name subset, then these are marked as "geneset level annotations"
    and tacked onto the GeneSetDb@table data.table. I put this in to pave the
    way for simple support of "tags" per geneset, but there are other uses
    as well.

CHANGES IN VERSION 0.6.x
------------------------

USER VISIBLE CHANGES

  o MSigDB version updated to 5.2 in 0.6.5
  o Interactive plots can be generated via iplot
  o single sample scoring methods broken out into more modular functions
  o svdScore added. Executes Jason's SVD scoring trick and also calculates
    "all the usual" prcomp stuff.
  o Adds a `[.GeneSetDb` method so we can subset the geneSets out of a GeneSetDb
    more easily. Would still like to have a fully operational subset(gdb, ...).

NEW FEATURES

  o Adds `svdGeneSetTest` method. A gene x sample expression matrix is
    transformed into a geneset x sampel matrix, and "normal" differential
    expression expression testing is run over the geneset scores.

CHANGES IN VERSION 0.4.x
------------------------

USER VISIBLE CHANGES

  o scoreSingleSamples now defualts to returning a melted data.frame of results
    (previously we returned a (list of) matrix of scores). `melted=FALSE`
    parameter was changed to `as.matrix=FALSE`

  o Internal MSigDB collections updated to v5.1 (as of v0.4.30).
    The major difference in v5.1 vs v5.0 seems to be an updated c7 (immunologic)
    collection, which comes on the heels of this paper:

      Compendium of Immune Signatures Identifies Conserved and Species-Specific
      Biology in Response to Inflammation
      http://www.cell.com/immunity/abstract/S1074-7613(15)00532-4

  o All methods now default to returning data.frame(s) instead of data.table(s)
    data.table is still used internally, however I understand that some people
    don't want to have them returned into their workspace.

    A global option (multiGSEA.df.return) is available for you to tweak this
    behavior, ie. set `options(multiGSEA.df.return='data.tabe')` to have
    `geneSets(gdb)` return a data.table instead of a data.frame.

  o Changed geneSetFeatureStatistics to geneSetsStats

NEW FEATURES

  o `use.treat` parameter has been added to relevant places to run the internal
    differential testing via the "treat" framework made available in limma and
    edgeR via the `treat` and `glmTreat` "pipelines," respectively. By default,
    all pipelines do not use the treat framework (but perhaps they should).

  o goseq, fry, and romer testing methods have been added.

  o Support for DGEList expression input fully baked in. Previously, a DGEList
    was shot through voom to be used internally. Internal differential gene
    expression statistics on DGELists use edgeR's quasi-likelihood framework.
    See `?edgeR::glmQLFit` for more information, and particulary reference
    Aaron Lun's tutorial on using this approach here:

      It's DE-licious: a recipe for differential expression analyses of RNA-seq
      experiments using quasi-likelihood methods in edgeR
      http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf

  o Adds support for PANTHER pathway (and GOSLIM) genesets from PANTHER.db
    package via the `getPantherGeneSetDb` function.

  o Added geneSetFeatureStats to return the logFC (and membership) information
    for a geneset after a multiGSEA run.

  o Added constructors and `as()` coercion functions to create GeneSetDb's
    from GeneSetCollection(s) and data.frames (and vice versa).

  o `scoreSingleSamples` is another wrapper function which calls several single
    sample geneset scoring algorithms, including the ones provided by the
    GSVA package (such as plage and ssGSEA)

  o Adds expression-utils.R for CPM/RPKM and meltx. Also has methods to
    facilitate "universal access" to pData, fData, and expression data from
    the different often-used expression containers
    (ExpressionSet, EList, DGEList, SummarizedExperiment)

  o MSigDB definitions were updated to v5.0, which adds a new "hallmark (h)"
    gene set. A call to `getMSigDBset` without a version parameter will return
    the v5.0 data.

    Note that MSigDB provides the entrez IDs for each geneset as human genes.
    The mouse version of these gene sets was constructing by mapping the
    human entrez ID's to their mouse orthologs via `igis::orthologs`. When
    this mapping returns multiple mouse IDs for a single human ID, all of
    the mouse IDs are kept in the mouse geneset.

    To get the previous v4.0 genesets (provided by WEHI), you would simply
    specify version='v4.0' in the getMSigDBSet like so:

      gdb5 <- getMSigDBsetc(c('c2', 'c7'), species='mouse')
      gdb4 <- getMSigDBsetc(c('c2', 'c7'), species='mouse', version='v4.0')


BUG FIXES

  o do.hyperGeometricTest was failing when the expression object `x` was a
    matrix.


CHANGES IN VERSION 0.1
----------------------


R/plots.R
## TODO: Add DOSE::gseaplot

##' Static versions of the iplot functionality
##'
##' @description
##' You can create static versions of the iplot functionality (density only).
##' This also provides a \code{type='barcode'} functionality that draws
##' limma style barcode plots.
##'
##' Note that the limma barcode plot is a direct copy of limma's functionality
##' into here, tweaked ever so slightly so that it's easy to pass GeneSetDb
##' information down through to it.
##'
##' @rdname plot
setMethod("plot", 'MultiGSEAResult',
function(x, y, j, value=c('logFC', 't'), type=c('density', 'barcode', 'mini'),
         main=NULL, ...) {
  value <- match.arg(value)
  type <- match.arg(type)
  if (!is.character(y) && !is.character(j)) {
    stop("y,j should be characters to identifiy a geneset by collection,name")
  }
  if (missing(main)) {
    main <- paste(y, j, sep=':')
  }

  fn <- getFunction(sprintf('plot.gsea.%s', type))
  fn(x, y, j, value, main, ...)
})

plot.gsea.density <- function(x, y, j, value, main, bg.density=NULL,
                              xlim=base::range(c(bg.density$x, gs.density$x)),
                              ylim=c(0, base::max(c(gs.density$y, bg.density$y)) + 0.1),
                              jitter.sig.higher=FALSE, with.rug=TRUE,
                              gs.name=j, lwd=3, ...) {
  lfc <- logFC(x, .external=FALSE)
  if (!is(bg.density, 'density')) {
    bg.density <- density(lfc[[value]], na.rm=TRUE)
  }
  gs.fids <- featureIds(x@gsd, y, j, 'x.id')
  gs.lfc <- lfc[gs.fids, on='featureId']
  gs.lfc  <- gs.lfc[!is.na(featureId)]
  gs.n <- nrow(gs.lfc)
  gs.lfc$xval <- gs.lfc[[value]] ##jitter(gs.lfc[[value]])

  ## This is probably a better default when looking at one plot at a time.
  ## The default uses the same xlim for all plots from a result, which is
  ## probably what we want when running through rmd.multiGSEA for quicker
  ## comparisons.
  ## if (missing(xlim)) {
  ##   qtl <- max(abs(quantile(lfc[[value]], c(0.01, 0.99))),
  ##              abs(gs.lfc[[value]]))
  ##   xlim <- c(-qtl, qtl)
  ## }

  if (jitter.sig.higher) {
    gs.lfc$yval <- ifelse(gs.lfc$significant, rep(0.05, gs.n), rep(0.02, gs.n))
  } else {
    gs.lfc$yval <- rep(0.02, gs.n)
  }
  gs.lfc[, yval := pmax(jitter(yval, amount=0.02), 0)]

  if (gs.n >= 3) {
    gs.density <- density(gs.lfc[[value]])
  } else {
    gs.density <- list(x=numeric(), y=numeric())
  }

  plot(bg.density, main=main, sub=sprintf('(%d features)', gs.n), xlab=value,
       ylim=ylim, xlim=xlim, lwd=lwd)
  if (is(gs.density, 'density')) {
    lines(gs.density, col='red', lwd=lwd)
  }

  if (with.rug) {
    with(subset(gs.lfc, !significant), {
      points(xval, yval, pch=16, col='#FF0000AA', cex=0.5)
    })
    with(subset(gs.lfc, significant), {
      points(xval, yval, pch=16, col='#FF0000AA', cex=0.8)
      points(xval, yval, pch=1L, col='#000000AA', cex=0.8, lwd=2)
    })
  }

  legend('topright', legend=c("all genes", gs.name), text.col=c('black', 'red'))

  attr(gs.lfc, 'plot.attrs') <- list(xlim=xlim, ylim=ylim)
  invisible(gs.lfc)
}

plot.gsea.barcode <- function(x, y, j, value, main, ...) {
  lfc <- logFC(x, .external=FALSE)

  gs.fids <- featureIds(x@gsd, y, j, 'x.id')
  idx <- lfc[gs.fids, which=TRUE, on='featureId']
  idx <- idx[!is.na(idx)]
  gs.n <- length(idx)

  main <- sprintf("%s (%s values)", main, value)
  plot.barcode(lfc[[value]], idx, weights.label=value, main=main)
}

plot.gsea.mini <- function(x, y, j, value, main, ...) {
  stop("plot.gsea.mini not yet implemented")
}


##' limma's barcodeplot of one or two genesets (with enrichment)
##'
##'
##' This was imported from the barcode plot in limma_3.22.1 (the function was
##' annotated with a last update date of 20 October 2008)
plot.barcode <- function(statistics, index=NULL, index2=NULL, gene.weights=NULL,
                         weights.label="Weight", labels=c("Up", "Down"),
                         quantiles=c(-1, 1), col.bars=NULL, worm=TRUE,
                         span.worm=0.45, ...) {
  ##	Check statistics
  if (!is.vector(statistics, mode="numeric")) {
    stop("statistics should be a numeric vector")
  }
  nstat <- length(statistics)

  ##	Check index
  if (is.null(index)) {
    if (is.null(gene.weights)) {
      stop("Must specify at least one of index or gene.weights")
    } else {
      if (length(gene.weights) == nstat) {
        index <- rep_len(TRUE, nstat)
        index2 <- NULL
      } else {
        stop("No index and length(gene.weights) doesn't equal length(statistics)")
      }
    }
  } else {
    if (any(is.na(index)))
      stop("Need to provide index without NAs")
    if (is.logical(index))
      if(length(index) != nstat)
        stop("Length of index disagrees with statistics")
    if (length(index) > nstat)
      stop("Length of index disagrees with statistics")
  }

  ##	Check index2
  if (!is.null(index2)) {
    if (!is.null(gene.weights))
      warning("gene.weights ignored")
    gene.weights <- statistics
    gene.weights[] <- 0
    gene.weights[index] <- 1
    gene.weights[index2] <- -1
    index <- rep_len(TRUE, nstat)
    index2 <- NULL
  }

  ##	Check gene.weights
  if (!is.null(gene.weights)){
    if (!is.vector(gene.weights, mode = "numeric"))
      stop("gene.weights should be a numeric vector")
    if (any(is.na(gene.weights)))
      stop("Need to provide gene.weights without NAs")
    if (all(gene.weights == 0))
      stop("gene.weights equal to zero: no selected genes to plot")
    if (length(gene.weights) != length(statistics[index]))
      stop("Length of gene.weights disagrees with size of set")

    one <- all(gene.weights >= 0) | all(gene.weights <= 0)

    if (one) {
      index2 <- NULL
      gene.weights1 <- rep_len(0, nstat)
      names(gene.weights1) <- names(statistics)
      gene.weights1[index] <- gene.weights

      index <- rep_len(FALSE, nstat)
      names(index) <- names(statistics)
      index[gene.weights1 != 0] <- TRUE

      gene.weights1 <- gene.weights1[gene.weights1 != 0]
      gene.weights <- gene.weights1
    } else {
      gene.weights12 <- rep_len(0, nstat)
      names(gene.weights12) <- names(statistics)
      gene.weights12[index] <- gene.weights

      index <- index2 <- rep_len(FALSE, nstat)
      names(index) <- names(index2) <- names(statistics)
      index[gene.weights12 > 0] <- TRUE
      index2[gene.weights12 < 0] <- TRUE

      gene.weights1 <- gene.weights12[gene.weights12 > 0]
      gene.weights2 <- gene.weights12[gene.weights12 < 0]

      gene.weights <- gene.weights1
    }
  }

  ##	Are there up and down sets?
  TWO <- !is.null(index2)

  ##	Convert indexes to logical and add gene.weights
  set1 <- set2 <- data.frame(idx = rep.int(FALSE,nstat), weight = NA, wt = NA)
  rownames(set1) <- rownames(set2) <- names(statistics)
  set1$idx[index] <- TRUE
  if (TWO) set2$idx[index2] <- TRUE

  if (length(gene.weights)) {
    set1$weight <- 0
    set1$weight[index]<- gene.weights
    set1$wt <- abs(set1$weight)/sum(abs(set1$weight))
    if (TWO) {
      set2$weight <- 0
      set2$weight[index2] <- gene.weights2

      SUM <- sum(abs(set1$weight), abs(set2$weight))
      set1$wt <- abs(set1$weight)/SUM
      set2$wt <- abs(set2$weight)/SUM

    }
  }

  ## TODO: enable sorting of statistics from low to high
  ## Sort statistics and indexes
  ostat <- order(statistics, na.last = TRUE, decreasing=TRUE)
  statistics <- statistics[ostat]
  set1 <- set1[ostat,]

  if (TWO) set2 <- set2[ostat,]

  ##	Trim off missing values
  n <- sum(!is.na(statistics))
  if (n==0L) {
    message("No valid statistics")
    return(invisible())
  }
  if (n < nstat) {
    statistics <- statistics[1:n]
    set1 <- set1[1:n,]
    if (TWO) set2 <- set2[1:n,]
  }

  ##	Convert indexes to integer
  r <- which(set1$idx)

  if (TWO) {
    r2 <- which(set2$idx)
    if(!length(r2)) TWO <- FALSE
  }

  #	Check there is something to plot
  if (!length(r)) {
    if(TWO) {
      r <- r2
      set1 <- set2
      TWO <- FALSE

    } else {
      message("No selected genes to plot")
      return(invisible())
    }
  }

  ##	Are there unequal weights?
  WTS <- FALSE
  wt1 <- set1$wt[r]
  len.up <- 1

  if (all(!is.na(wt1))) {
    len.up <- set1$weight[r]/max(abs(set1$weight[r]))

    anydifferent <- function(x) {
      if(length(x) < 2) return(FALSE)
      r <- range(x)
      (r[2] > r[1])
    }

    if (!TWO) if (anydifferent(wt1)) WTS <- TRUE

    if (TWO) {
      wt12 <- c(set1$wt[r], abs(set2$wt[r2]))
      if(anydifferent(wt12)) WTS <- TRUE

      max.wt <- max(set1$wt[r], set2$wt[r2])
      len.up <- set1$wt[r]/max.wt
      len.down <- set2$wt[r2]/max.wt
    }
  }

  pos.dir <- all(len.up > 0)

  #	Plot setting
  if (WTS) shift <- 0.1 else shift <- 0

  #	Check other arguments
  quantiles <- sort(quantiles)

  #	color settting
  ALP <- 0.4

  if (is.null(col.bars)) {
    if (TWO) {
      col.bars <- c("red", "blue")
      if(WTS) col.bars.alpha <- c(rgb(1,0,0,alpha=ALP), rgb(0,0,1,alpha=ALP))
      else col.bars.alpha <- col.bars

    } else {
      col.bars <- "black"
      if(WTS) col.bars.alpha <- rgb(0,0,0,alpha=ALP)
      else col.bars.alpha <- col.bars
    }
  } else {
    if (TWO) {
      if(length(col.bars) == 1) col.bars <- rep(col.bars, 2)
      RGB <- col2rgb(col.bars)/255
      red <- RGB[1,1]
      green <- RGB[2,1]
      blue <- RGB[3,1]
      red2 <- RGB[1,2]
      green2 <- RGB[2,2]
      blue2 <- RGB[3,2]
      if (WTS) {
        col.bars.alpha <- c(rgb(red, green, blue, alpha=ALP),
                            rgb(red2, green2, blue2, alpha=ALP))
      } else {
        col.bars.alpha <- col.bars
      }
    } else {
      RGB <- col2rgb(col.bars)/255
      red <- RGB[1,1]
      green <- RGB[2,1]
      blue <- RGB[3,1]
      if(WTS) col.bars.alpha <- rgb(red, green, blue, alpha=ALP)
      else col.bars.alpha <- col.bars
    }
  }

  ## TODO: Enable different labeling of of statistics
  ## worm plot setting
  ylim.worm <- ylim <- c(-1, 1)
  ylab.worm <- ""
  xlab.worm <- "statistics"

  if (!TWO) ylim.worm <- c(0, 1)

  if (worm) {
    ylim.worm <- c(-2.1, 2.1)
    if (!TWO) ylim.worm <- c(0, 2.1)
  }

  ylim[2] <- ylim[2] + 0.5
  if (TWO) ylim[1] <- ylim[1] - 0.5

  if (TWO) {
    plot(1:n, xlim=c(0,n), ylim=c(ylim.worm[1]-shift, ylim.worm[2]+shift),
         type="n", axes=FALSE, xlab=xlab.worm, ylab=ylab.worm, ...)
  }
  if (!TWO) {
    plot(1:n, xlim=c(0,n),
         ylim=c(ylim.worm[1]-shift*(!pos.dir), ylim.worm[2]+shift*pos.dir),
         type="n", axes=FALSE, xlab=xlab.worm,ylab=ylab.worm, ...)
  }

  npos <- sum(statistics > quantiles[2])
  nneg <- sum(statistics < quantiles[1])

  lwd <- 50/length(r)
  lwd <- min(1.9, lwd)
  lwd <- max(0.2, lwd)

  if (TWO){
    lwd2 <- 50/length(r2)
    lwd2 <- min(1.9, lwd2)
    lwd2 <- max(0.2, lwd2)

    lwd <- lwd2 <- min(lwd, lwd2)
  }

  barlim <- ylim[2] - c(1.5, 0.5)

  if (!pos.dir) {
    rect.yb <- 0.5
    rect.yt <- 1
    rect(npos + 0.5, rect.yb, n - nneg + 0.5, rect.yt,
         col = "lightgray", border = NA)
    if (npos)
      rect(0.5, rect.yb, npos + 0.5, rect.yt, col = "pink",
           border = NA)
    if (nneg)
      rect(n - nneg + 0.5, rect.yb, n + 0.5, rect.yt, col = "lightblue",
           border = NA)

    segments(r, barlim[2]/2, r, barlim[2], lwd = lwd, col = col.bars.alpha[1])
    segments(r, barlim[2]/2-shift, r, barlim[2]/2*(1+len.up)-shift, lwd = lwd,
             col = col.bars[1])
  }

  if (pos.dir) {
    rect.yb <- -0.5
    if (!TWO) rect.yb <- 0
    rect.yt <- 0.5

    rect(npos + 0.5, rect.yb, n - nneg + 0.5, rect.yt, col = "lightgray",
         border = NA)
    if (npos)
      rect(0.5, rect.yb, npos + 0.5, rect.yt, col = "pink", border = NA)
    if (nneg)
      rect(n - nneg + 0.5, rect.yb, n + 0.5, rect.yt, col = "lightblue",
           border = NA)

    segments(r, barlim[1], r, barlim[2]/2, lwd = lwd, col = col.bars.alpha[1])
    segments(r, barlim[2]/2+shift, r, barlim[2]/2*(1+len.up)+shift, lwd = lwd,
             col = col.bars[1])
  }

  if (TWO) {
    barlim2 <- ylim[1] + c(0.5, 1.5)
    segments(r2, barlim2[1]/2, r2, barlim2[2], lwd = lwd2,
             col = col.bars.alpha[2])
    segments(r2, barlim2[1]/2*(1+len.down)-shift, r2, barlim2[1]/2-shift,
             lwd = lwd2, col = col.bars[2])
  }

  lab.at <- 0
  if (!TWO) lab.at <- 0.5
  axis(side = 2, at = lab.at, padj = 3.8, cex.axis = 0.85, labels = labels[1],
       tick = FALSE)
  axis(side = 4, at = lab.at, padj = -3.8, cex.axis = 0.85, labels = labels[2],
       tick = FALSE)

  ## label statistics on x axis
  prob <- (10:0)/10
  axis(at = seq(1,n,len=11), side = 1, cex.axis = 0.7, las = 2,
       labels = format(quantile(statistics, p = prob), digits = 1))

  ## create worm
  if (worm) {
    # rescale x to new range
    rescale <- function(x, newrange, oldrange=range(x)) {
      newrange[1] + (x-oldrange[1]) / (oldrange[2]-oldrange[1]) * (newrange[2] - newrange[1])
    }

    # calculate enrichment
    if (!WTS) {
      ave.enrich1 <- length(r)/n
      worm1 <- tricubeMovingAverage(set1$idx, span = span.worm)/ave.enrich1
      if(TWO) {
        ave.enrich2 <- length(r2)/n
        worm2 <- tricubeMovingAverage(-set2$idx, span = span.worm)/ave.enrich2
      }
    }

    ## calculate weighted enrichment
    if (WTS) {
      ave.enrich1 <- mean(set1$wt)
      worm1 <- tricubeMovingAverage(set1$wt, span = span.worm)/ave.enrich1

      if (TWO) {
        ave.enrich2 <- mean(set2$wt)
        worm2 <- tricubeMovingAverage(-set2$wt, span = span.worm)/ave.enrich2
      }
    }

    # rescale worm
    max.worm1 <- max(worm1)
    r.worm1 <- c(0,max.worm1)
    worm1.scale <- rescale(worm1,
                           newrange=c(1.1+shift*pos.dir,2.1+shift*pos.dir),
                           oldrange=r.worm1)

    if (TWO) {
      min.worm2 <- min(worm2)
      r.worm2 <- c(min.worm2,0)
      worm2.scale <- rescale(worm2,
                             newrange=c(-2.1-shift,-1.1-shift),
                             oldrange=r.worm2)
    }

    ## plot worms
    if (!TWO) {
      lines(x = 1:n, y = worm1.scale, col = col.bars[1], lwd = 2)
      abline(h = rescale(1,newrange=c(1.1+shift*pos.dir,2.1+shift*pos.dir),oldrange=r.worm1),
             lty=2)
      axis(side = 2,
           at = c(1.1+shift*pos.dir, 2.1+shift*pos.dir),
           cex.axis = 0.8,
           labels = c(0, format(max.worm1, digits = 2)))
      axis(side = 2,
           labels = "Enrichment",
           at = 1.6+shift*pos.dir, padj = -0.6,
           tick = FALSE,
           cex.axis = 0.8)
    }

    if (TWO) {
      lines(x = 1:n, y = worm1.scale, col = col.bars[1], lwd = 2)
      abline(h = rescale(1,newrange=c(1.1+shift,2.1+shift),oldrange=r.worm1),
             lty=2)
      lines(x = 1:n, y = worm2.scale, col = col.bars[2], lwd = 2)
      abline(h = rescale(-1,newrange=c(-2.1-shift,-1.1-shift),oldrange=r.worm2),
             lty=2)

      axis(side = 2,
           at = c(1.1+shift, 2.1+shift),
           cex.axis = 0.7,
           labels = c(0, format(max.worm1, digits = 2)))
      axis(side = 2,
           at = c(-1.1-shift, -2.1-shift),
           cex.axis = 0.7,
           labels = c(0, format(-min.worm2, digits = 2)))
      axis(side = 2,
           labels = "Enrichment",
           at = 1.6+shift,
           tick = FALSE,
           padj = -0.6,
           cex.axis = 0.7)
      axis(side = 2,
           labels = "Enrichment",
           at = -1.6-shift,
           tick = FALSE,
           padj = -0.6,
           cex.axis = 0.7)
    }
  }

  ## add gene.weights axis
  if (WTS){
    if (!TWO){
      if (pos.dir) {
        axis(side = 2,
             at = c(0.5+shift, 1+shift),
             cex.axis = 0.48,
             padj = 1.6,
             labels = c(0, format(max(set1$weight[r]), digits = 2)))
        axis(side = 2,
             labels = weights.label[1],
             at = 0.75+shift,
             padj = 1,
             tick = FALSE,
             cex.axis = 0.5)
      }

      if (!pos.dir){
        axis(side = 2,
             at = c(0-shift, 0.5-shift),
             cex.axis = 0.48,
             padj = 1.6,
             labels = c(format(min(set1$weight[r]), digits = 2), 0))
        axis(side = 2,
             labels = weights.label[1],
             at = 0.25-shift,
             padj = 1,
             tick = FALSE,
             cex.axis = 0.5)
      }
    }

    if (TWO){
      max.weight <- max(set1$weight[r], abs(set2$weight[r2]))
      axis(side = 2,
           at = c(0.5+shift, 1+shift),
           cex.axis = 0.43,
           labels = c(0, format(max.weight, digits = 2, scientific = FALSE)),
           padj = 1.6)
      axis(side = 2,
           labels = weights.label[1],
           at = 0.75+shift,
           padj = 1,
           tick = FALSE,
           cex.axis = 0.46)
      axis(side = 2,
           at = c(-0.5-shift, -1-shift),
           cex.axis = 0.43,
           labels = c(0, format(-max.weight, digits = 2, scientific = FALSE)),
           padj = 1.6)
      axis(side = 2,
           labels = weights.label[1],
           at = -0.75-shift,
           padj = 1,
           tick = FALSE,
           cex.axis = 0.46)
    }
  }

  invisible()
}
